### PR TITLE
fix #3688, Optimization of table building statements

### LINKED ISSUE #3688: [Improvement][SQL optimization] Optimization of table building statements
1. When executing MySQL table creation statement, it will report that the index length is not enough.
this will happen: Index column size too large. The maximum column size is 767 bytes.
2. The length of the t_ds_task_instance.app_link is not enough. When hive SQL is complex and yarn will start multiple applications

### PR DESCRIPTION
**Pull request by JinyLeeChina**
1. When executing MySQL table creation statement, it will report that the index length is not enough.
this will happen: Index column size too large. The maximum column size is 767 bytes.
2. The length of the t_ds_task_instance.app_link is not enough. When hive SQL is complex and yarn will start multiple applications

### CODE REVIEW

File: sql/dolphinscheduler_mysql.sql [hunk 1]
```diff
@@ -248,7 +248,7 @@ CREATE TABLE `QRTZ_TRIGGERS` (
   KEY `IDX_QRTZ_T_NFT_ST_MISFIRE` (`SCHED_NAME`,`MISFIRE_INSTR`,`NEXT_FIRE_TIME`,`TRIGGER_STATE`),
   KEY `IDX_QRTZ_T_NFT_ST_MISFIRE_GRP` (`SCHED_NAME`,`MISFIRE_INSTR`,`NEXT_FIRE_TIME`,`TRIGGER_GROUP`,`TRIGGER_STATE`),
   CONSTRAINT `QRTZ_TRIGGERS_ibfk_1` FOREIGN KEY (`SCHED_NAME`, `JOB_NAME`, `JOB_GROUP`) REFERENCES `QRTZ_JOB_DETAILS` (`SCHED_NAME`, `JOB_NAME`, `JOB_GROUP`)
-) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
+) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```
Comment (yangyichao-mango) [inline sql/dolphinscheduler_mysql.sql:251]: Why need to use `utf8` to replace `utd8mb4`?

**Review by yangyichao-mango — DISMISSED** [review]

File: sql/dolphinscheduler_mysql.sql (hunk 1 above)
Comment (felix-thinkingdata) [reply sql/dolphinscheduler_mysql.sql:251]: https://www.cnblogs.com/cuiqq/p/11045487.html

I read this post and didn't quite understand the need to change the coding format.

Comment (JinyLeeChina) [reply sql/dolphinscheduler_mysql.sql:251]: I checked the qrtz related tables and will not write 4-byte Unicode. Utf8mb4 encoding will cause the index of qrtz related tables to exceed the default value of MySQL, which will cause many users to fail to create tables. 
image

Comment (yangyichao-mango) [reply sql/dolphinscheduler_mysql.sql:251]: > I checked the qrtz related tables and will not write 4-byte Unicode. Utf8mb4 encoding will cause the index of qrtz related tables to exceed the default value of MySQL, which will cause many users to fail to create tables. 
> 
> image
> 
> 

Which version of your mysql?

Comment (JinyLeeChina) [reply sql/dolphinscheduler_mysql.sql:251]: Version 5.6, I think there are many users of this version of MySQL, we should be compatible

Comment (lgcareer) [reply sql/dolphinscheduler_mysql.sql:251]: Hi,we should discuss first,the utf8mb4 is suggested to use in mysql.You can see the article below.
https://medium.com/@adamhooper/in-mysql-never-use-utf8-use-utf8mb4-11761243e434
I saw the error you pointed `Index column size too large. The maximum column size is 767 bytes.`The another article give a method to resolve it.
https://www.jianshu.com/p/5c0cce24189b

Comment (lgcareer) [reply sql/dolphinscheduler_mysql.sql:251]: I will send a email in the dev list,we can discuss it which charset we need to choose.

**Comment by lgcareer** [comment]
@JinyLeeChina @yangyichao-mango @all
I sent a email to discuss that which charst should we choose in mysql,We can discuss in the email thread below.
https://lists.apache.org/thread.html/r6a1ebb8bf5767571cdabfc625cad62611f9b04d9c06671668e9ebef7%40%3Cdev.dolphinscheduler.apache.org%3E

**Comment by dailidong** [comment]
> @JinyLeeChina @yangyichao-mango @ALL
> I sent a email to discuss that which charst should we choose in mysql,We can discuss in the email thread below.
> https://lists.apache.org/thread.html/r6a1ebb8bf5767571cdabfc625cad62611f9b04d9c06671668e9ebef7%40%3Cdev.dolphinscheduler.apache.org%3E

the conclusion is to use utf8

**Review by dailidong — APPROVED** [review]
+1
good job

**Pull request merged**